### ISSUE TITLE
redo_pyc is too slow

### ISSUE THREAD

**Issue by k4ml** [buildout/buildout#157]
Installing large package such as Django on EC2 micro instance took a
very long time, 8-9 minutes with 99% cpu usage. Initially, I taught it
caused by setuptools analyzing the packages to figure out it zip_safe
or not [1]. But after looking at this closely, that's not the case.
Analyzing the egg only took few seconds and can be negligible to the
total time it took to install the whole package. I have also test by
adding zip_safe=False to django's setup.py and didn't see any drastic
improvement to the time taken to install it.

I test by using easy_install directly and it took around 1-2 minutes
to finish so it mean the other 8 minutes being spent in buildout
itself rather than in setuptools/easy_install. The install process
basically went like this:-

...
Writing /tmp/easy_install-GwjQPW/django-master/setup.cfg
Running django-master/setup.py -q bdist_egg --dist-dir
/tmp/easy_install-GwjQPW/django-mas
ter/egg-dist-tmp-Yk_MYR
warning: no previously-included files matching '**pycache**' found
under directory '_'
warning: no previously-included files matching '_.py[co]' found under
directory '*'
...
...
<LONG GAP HERE ...>
Got Django 1.7.
Picked: Django = 1.7
Generated script '/home/kamal/test_buildout/bin/django-admin.py'.
Generated interpreter '/home/kamal/test_buildout/bin/python'.

Stepping through the code, I figure out the LONG GAP starting after:-

```
                dists = self._call_easy_install(
                    dist.location, ws, self._dest, dist)
```

in line 531 of zc/buildout/easy_install.py. Next after this line is:-

```
                for dist in dists:
                    redo_pyc(dist.location)
```

Commenting this function call I manage to cut down the installation
time to 2m30s. So what the purpose of this function ? Skipping it seem
to be fine, I can import the package without any error. My
buildout.cfg:-

[buildout]
parts = base

[base]
recipe = zc.recipe.egg
eggs =
        Django
interpreter = python

The only reference to redo_pyc I found is
http://www.makina-corpus.org/blog/minitage-10-out which just saying
redo_pyc to be somehow slow.

**Comment by leorochael** [comment]
That's exactly right, the `redo_pyc` call is there so that the `.pyc` or `.pyo` files get (re)compiled. This makes the initial import of the affected modules faster, but removing this call doesn't impact the usability of the installed modules, since Python either builds the `.pyc` files itself on first run if they're missing, or uses the ones that are already there.

The worst that can happen is if the installed packages come already with their own badly built .pyc files that would report the paths incorrectly in Python tracebacks, but this doesn't tend to affect the actual usability of the packages.

In any case, the `redo_pyc()` call is there by design, and I suggest closing this issue, but if you feel like this should be addressed, a Pull Request making the `redo_pyc()` call optional could be considered.

**Comment by mgedmin** [comment]
Many packages use `os.path.dirname(__file__)` to locate resources (e.g. templates).  Having incorrect paths in pyc files would break those packages.

**Comment by leorochael** [comment]
@mgedmin right, then we could consider a PR that would replace `redo_pyc()` with a `redo_or_remove_pyc()` that could optionally remove `.pyc` or `.pyo` files IFF there are equivalent `.py` files present (after all, there could be odd proprietary packages that distributes only .pyc files).

Another possibility for someone who is interested in speeding up this step would be to spawn many subprocesses (e.g. one per CPU) to compile `.pyc` files in parallel.

In any case, I think this issue could be closed, or at least reclassified as enhancement.

**Comment by mauritsvanrees** [comment]
I have removed the `redo_pyc` function in PR #702, released in `zc.buildout 4.1.7.  It's current implementation was useless because it did not do anything on Python 3.
